### PR TITLE
fix(relay): Retry upon TCP connection error

### PR DESCRIPTION
**Pull request by Cardosaum**
Solves #842

### CODE REVIEW

File: bonsai/ethereum-relay/src/uploader/pending_proofs/pending_proof_request_future.rs [hunk 1]
```diff
@@ -98,22 +99,44 @@ impl Future for PendingProofRequest {
...
-                            if receipt_response.status == "SUCCEEDED" {
-                                return Poll::Ready(Ok(this.pending_proof_id.clone()));
+                            match (
+                                receipt_response.status.as_str(),
+                                receipt_response.receipt_url.is_some(),
+                            ) {
+                                ("SUCCEEDED", true) => {
+                                    return Poll::Ready(Ok(this.pending_proof_id.clone()))
+                                }
+                                ("SUCCEEDED", false) => {
```
Comment (mothran) [inline bonsai/ethereum-relay/src/uploader/pending_proofs/pending_proof_request_future.rs:109]: Yeah if this happens its a hard error and bonsai has something very wrong with it haha.

File: bonsai/ethereum-relay/Cargo.toml [hunk 2]
```diff
@@ -51,6 +52,7 @@ validator = { version = "0.16", features = ["derive"] }
 bincode = "1"
 bytemuck = "1.13"
 risc0-zkvm-methods = { path = "../../risc0/zkvm/methods", default-features = false }
+rstest = "0.18"
```
Comment (mothran) [inline bonsai/ethereum-relay/Cargo.toml:55]: https://docs.rs/rstest/latest/rstest/ interesting! TIL, @flaub this looks like a interesting library for testing.

**Review by mothran — APPROVED** [review]
LGTM

File: bonsai/ethereum-relay/Cargo.toml (hunk 2 above)
Comment (Cardosaum) [reply bonsai/ethereum-relay/Cargo.toml:55]: You can see it's usage [here] for example.

[here]: https://github.com/risc0/risc0/pull/903/files#diff-e45dd1869f09cca9bbdd8d39cd54d53427049fc3c256e3265f7497921ca5397bR88-R107

**Pull request merged**